doc/source/reference/c-api/array.rst

```diff
@@ -1555,8 +1555,398 @@ For all of these macros *arr* must be an instance of a (subclass of)
+   element of a single array.
+
+   Currently this is used for array clearing, via the ``NPY_DT_get_clear_loop``
+   DType API hook, and zero-filling, via the ``NPY_DT_get_fill_zero_loop``
+   DType API hook.  These are most useful for handling arrays storing embedded
+   references to python objects or heap-allocated data.
+
+   The *descr* is the descriptor for the array, *data* is a pointer to the array
+   buffer, *size* is the 1D size of the array buffer, *stride* is the stride,
+   and *auxdata* is optional extra data for the loop.
+
+   The *traverse_context* is passed in because we may need to pass in
+   Interpreter state or similar in the future, but we don't want to pass in a
+   full context (with pointers to dtypes, method, caller which all make no sense
+   for a traverse function). We assume for now that this context can be just
+   passed through in the future (for structured dtypes).
+
+.. c:type:: int (PyArrayMethod_GetTraverseLoop)( \
+                void *traverse_context, PyArray_Descr *descr, \
+                int aligned, npy_intp fixed_stride, \
+                PyArrayMethod_TraverseLoop **out_loop, NpyAuxData **out_auxdata, \
+                NPY_ARRAYMETHOD_FLAGS *flags)
+
+   Simplified get_loop function specific to dtype traversal
+
+   It should set the flags needed for the traversal loop and set *out_loop* to the
+   loop function, which must be a valid ``PyArrayMethod_TraverseLoop``
+   pointer. Currently this is used for zero-filling and clearing arrays storing
+   embedded references.
+
+API Functions and Typedefs
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+These functions are part of the main numpy array API and were added along
+with the rest of the ArrayMethod API.
+
+.. c:function::  int PyUFunc_AddLoopFromSpec( \
+                         PyObject *ufunc, PyArrayMethod_Spec *spec)
+
+   Add loop directly to a ufunc from a given ArrayMethod spec.
+   the main ufunc registration function.  This adds a new implementation/loop
+   to a ufunc.  It replaces `PyUFunc_RegisterLoopForType`.
+
+.. c:function:: int PyUFunc_AddPromoter( \
+                        PyObject *ufunc, PyObject *DType_tuple, PyObject *promoter)
+
+   Note that currently the output dtypes are always ``NULL`` unless they are
+   also part of the signature. This is an implementation detail and could
+   change in the future. However, in general promoters should not have a
+   need for output dtypes.
+   Register a new promoter for a ufunc. The first argument is the ufunc to
+   register the promoter with. The second argument is a Python tuple containing
+   DTypes or None matching the number of inputs and outputs for the ufuncs. The
+   last argument is a promoter is a function stored in a PyCapsule.  It is
+   passed the operation and requested DType signatures and can mutate it to
+   attempt a new search for a matching loop/promoter.
+
+.. c:type:: int (PyArrayMethod_PromoterFunction)(PyObject *ufunc, \
```

Comment:
Isn't this a `c:function::`?

Reply:
I looked through a bit and it seemed right?  This is a function type definition (user defines something with that type), while `c:function::` seems correctly used for functions the user would call.

Reply:
Yup, see https://www.sphinx-doc.org/en/master/usage/domains/c.html#directive-c-type. All the ones specified with `type` are function typedefs.

Reply:
Ah, sorry that i missed that!